instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java

```diff
@@ -54,189 +42,63 @@
-  private static final String DEFAULT_TRACE_ANNOTATED_METHODS_EXCLUDE = null;
-
-  public static final String SQL_NORMALIZER_ENABLED = "sql.normalizer.enabled";
-  public static final boolean DEFAULT_SQL_NORMALIZER_ENABLED = true;
-
-  private final String exporterJar;
-  private final String exporter;
-  private final List<String> propagators;
-  private final boolean traceEnabled;
-  private final boolean integrationsEnabled;
-  private final List<String> excludedClasses;
-  private final Integer scopeDepthLimit;
-  private final boolean runtimeContextFieldInjection;
-
-  private final String traceAnnotations;
-
-  private final String traceMethods;
-  private final String traceAnnotatedMethodsExclude;
-
-  private final boolean traceExecutorsAll;
-  private final List<String> traceExecutors;
-
-  private final boolean sqlNormalizerEnabled;
-
-  private final boolean kafkaClientPropagationEnabled;
-
-  private final boolean hystrixTagsEnabled;
-
-  private final Map<String, String> endpointPeerServiceMapping;
-
-  // Values from an optionally provided properties file
-  private static Properties propertiesFromConfigFile;
-
-  // Read order: System Properties -> Env Variables, [-> properties file], [-> default value]
-  // Visible for testing
-  Config() {
-    propertiesFromConfigFile = loadConfigurationFile();
-
-    propagators = getListSettingFromEnvironment(PROPAGATORS, null);
-    exporterJar = getSettingFromEnvironment(EXPORTER_JAR, null);
-    exporter = getSettingFromEnvironment(EXPORTER, "otlp");
-    traceEnabled = getBooleanSettingFromEnvironment(TRACE_ENABLED, DEFAULT_TRACE_ENABLED);
-    integrationsEnabled =
-        getBooleanSettingFromEnvironment(INTEGRATIONS_ENABLED, DEFAULT_INTEGRATIONS_ENABLED);
-
-    excludedClasses = getListSettingFromEnvironment(TRACE_CLASSES_EXCLUDE, null);
-
-    scopeDepthLimit =
-        getIntegerSettingFromEnvironment(SCOPE_DEPTH_LIMIT, DEFAULT_SCOPE_DEPTH_LIMIT);
-
-    runtimeContextFieldInjection =
-        getBooleanSettingFromEnvironment(
-            RUNTIME_CONTEXT_FIELD_INJECTION, DEFAULT_RUNTIME_CONTEXT_FIELD_INJECTION);
-
-    traceAnnotations = getSettingFromEnvironment(TRACE_ANNOTATIONS, DEFAULT_TRACE_ANNOTATIONS);
-
-    traceMethods = getSettingFromEnvironment(TRACE_METHODS, DEFAULT_TRACE_METHODS);
-    traceAnnotatedMethodsExclude =
-        getSettingFromEnvironment(
-            TRACE_ANNOTATED_METHODS_EXCLUDE, DEFAULT_TRACE_ANNOTATED_METHODS_EXCLUDE);
-
-    traceExecutorsAll =
-        getBooleanSettingFromEnvironment(TRACE_EXECUTORS_ALL, DEFAULT_TRACE_EXECUTORS_ALL);
+  private static final Config DEFAULT = Config.newBuilder().build();
 
-    traceExecutors = getListSettingFromEnvironment(TRACE_EXECUTORS, DEFAULT_TRACE_EXECUTORS);
+  // INSTANCE can never be null - muzzle instantiates instrumenters when it generates
+  // getInstrumentationMuzzle() and the Instrumenter.Default constructor uses Config
+  private static volatile Config INSTANCE = DEFAULT;
```

Comment:
`Config` becoming `volatile` is a surprisingly big change. It's going to make unit tests much easier where we currently use bytecode manipulation to turn `Config` from final to non-final so I'm generally happy about it though :)

We should confirm that we only call `Config.get` on cold paths, which I think should generally be the case as it's basically only  to initialize instrumentation. In Java, the performance of `Config.get().isFooEnabled()` is actually significantly different when Java can know a boolean is a constant and C1 compiler can just remove the branches with no de-opt mechanism at all. So when `.get` is volatile, we'd expect a `private static final boolean IS_FOO_ENABLED = Config.get().isFooEnabled()` to replace such code inline on a hot path where we would like fast enabling/disabling of code paths

Reply:
That's a good point -- I'll look through usages of `Config#get()`